main/core/src/eval/Evaluator.scala

```diff
@@ -598,221 +547,89 @@ case class Evaluator(
-      }.takeWhile { _ =>
-        // in case we have a collision, we try to unblock as fast as possible
-        taken < 1 || collisionsFree && taken < (effectiveThreadCount * 2)
-      }
-      evalLog.debug(s"Search for ${newInProgress.size} new dep-free tasks (with ${newSeen.size} duplicate drops) took ${System.currentTimeMillis() - scheduleStart} msec")
-
-      // update state
-      state.pending --= newInProgress
-      state.inProgress ++= newInProgress
+        implicit val ec = new ExecutionContext {
+          def execute(runnable: Runnable) = threadPool.submit(runnable)
+          def reportFailure(t: Throwable) {}
+        }
 
-      if (newInProgress.isEmpty) {
-        evalLog.debug(s"No new tasks to schedule (issuer: ${issuer})")
-      } else {
-        evalLog.debug(s"Scheduling ${newInProgress.size} new tasks (issuer: ${issuer}): ${newInProgress.map(t => printTerm(t._1)).mkString(", ")}")
-
-        // schedule for parallel execution
-        newInProgress.zipWithIndex.foreach {
-          case (curWork@(terminal, group), index) =>
-            val term = printTerm(terminal)
-            //
-            // Multithreading: Creating the Future
-            //
-            val workerFut: java.util.concurrent.Future[FutureResult] = completionService.submit { () =>
-              if (failFast && state.someTaskFailed.get()) {
-                // we do not start this tasks but instead return with aborted result
-                val newResults = group.map(_ -> Aborted)
-
-                evalLog.debug(s"Skipped evaluation (because of earlier failures): ${term}")
-                FutureResult(curWork, 0, Evaluated(newResults.toMap, Seq(), false))
-
-              } else {
-
-                val counterMsg = state.nextCounterMsg()
-
-                timeLog.debug(s"START [${counterMsg}] ${term}")
-                val startTime = System.currentTimeMillis()
-
-                val res = evaluateGroupCached(
-                  terminal,
-                  group,
-                  state.results.asScala,
-                  counterMsg,
-                  reporter,
-                  testReporter,
-                  logger
-                )
+        def label(x: Terminal) = x match {
+          case Left(x) => x.toString
+          case Right(Labelled(x, y)) => x.toString
+        }
 
-                val endTime = System.currentTimeMillis()
+        val groupGraph = interGroupDeps.items.map { case (k, vs) => (label(k._1), vs.map(t => label(t._1))) }.toMap
+        pprint.log(groupGraph, height=99999)
+        val terminals = sortedGroups.keys.toVector
+        //      println("INITIALIZING promises " + terminals.map(label).mkString(", "))
+        val promises = terminals
+          .map(k => (k, scala.concurrent.Promise[Any]))
+          .toMap
+
+        val taskFutures: Map[Terminal, Future[Any]] = promises.map { case (k, p) => (k, p.future) }
+
+        val results = mutable.LinkedHashMap.empty[Task[_], Result[(Any, Int)]]
```

Comment:
This guy should probably be made a ConcurrentHashMap, which would make all the `synchronized` blocks go away. I think this map is the only mutable data structure in the entire algorithm, except for the `workerCache` which we can also make concurrent